src/Ethereum/Transaction.h

```diff
@@ -9,54 +9,107 @@
 #include "Address.h"
 #include "../uint256.h"
 
+#include <memory>
+
 namespace TW::Ethereum {
 
-class Transaction {
+// Transactions can be:
+// - Legacy transactions: (pre-EIP2718)
+//  -- simple ETH transfer
+//  -- others with payload, function call, e.g. ERC20 transfer
+// - Enveloped transactions (EIP2718), with specific type, and specific transaction payload
+
+struct SignatureRSV {
+public:
+    uint256_t r;
+    uint256_t s;
+    uint256_t v;
+};
+
+/// Base class for all transactions.
+/// Legacy and various enveloped transactions derive from this.
+class TransactionBase {
 public:
     uint256_t nonce;
+    Data payload;
+
+public:
+    TransactionBase(const uint256_t& nonce, const Data& payload): nonce(nonce), payload(payload) {}
+    virtual ~TransactionBase() {}
+    // pre-sign encoded tx, for hashing
+    virtual Data hash(const uint256_t chainID) const = 0;
+    // encoded tx (signed)
+    virtual Data encoded(const SignatureRSV& signature, const uint256_t chainID) const = 0;
+
+protected:
+    TransactionBase() {}
+};
+
+/// Original transaction format, legacy as pre-EIP2718
+class TransactionLegacy: public TransactionBase {
```

Comment:
I prefer to use `Transaction` + other description, like code name or hard fork name